### PR TITLE
PTQ example for NeMo 2.0

### PR DESCRIPTION
**Pull request by Laplasjan107**
# What does this PR do ?

Example workflow for quantizing NeMo 2.0 checkpoint.

**Collection**: llm

# Known Issues 

* Observed accuracy regression on MMLU when exporting a non-quantized LLAMA3-7B: `0.654 -> 0.638`
* Half-precision with PP>1 does not work

# Usage
* You can potentially add a usage example below

```python
from nemo.collections.llm import quantization

quantization_config = quantization.QuantizationConfig()
export_config = quantization.ExportConfig('/tmp/trt_llama_engine')
quantizer = quantization.Quantizer(quantization_config, export_config)
model = quantization.load_with_modelopt_layer_spec('/path/to/nemo/checkpoint')
model = quantizer.quantize(model)
quantizer.export(model)
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: .github/workflows/cicd-main.yml [hunk 1]
```diff
@@ -5450,6 +5450,21 @@ jobs:
       SCRIPT: |
         bash tests/collections/llm/bitexact/mixtral/run.sh
 
+  L2_NeMo_2_PTQ_Llama2_FP8:
+    needs: [cicd-test-container-setup]
```
Comment (ko3n1g) [inline .github/workflows/cicd-main.yml:5454]: Just a friendly reminder to make sure it doesn't get lost once this PR is ready: this job name also needs to be added to `CICD_Nemo_Test` like the others

**Review by ko3n1g — CHANGES_REQUESTED** [review]

File: examples/llm/quantization/ptq.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    parser.add_argument("-ctp", "--calib_tp", type=int, default=1)
+    parser.add_argument("-cpp", "--calib_pp", type=int, default=1)
+    parser.add_argument("-tps", "--tensor_parallelism_size", type=int, default=1)
+    parser.add_argument("-pps", "--pipeline_parallelism_size", type=int, default=1)
+    parser.add_argument('-out', '--output_path', type=str, help='Path for the exported engine')
+    parser.add_argument(
+        '-algo',
+        '--algorithm',
+        type=str,
+        default="no_quant",
```
Comment (janekl) [inline examples/llm/quantization/ptq.py:39]: Let's maybe use "fp8" by default?

File: examples/llm/quantization/ptq.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    parser.add_argument(
+        '-calib_ds',
+        '--calibration_dataset',
+        default="cnn_dailymail",
+        choices=["wikitext", "cnn_dailymail"],
+        type=str,
+        help='Calibration dataset to be used',
+    )
+
+    return parser.parse_args(sys.argv[1:])
```
Comment (janekl) [inline examples/llm/quantization/ptq.py:65]: No need to pass `sys.argv[1:]` around

File: nemo/collections/llm/quantization/quantizer.py [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+        dtype = export_config["dtype"]
+
+        # Quantization sanity checks
+        assert algorithm is None or algorithm in QUANT_CFG_CHOICES, f"Unsupported quantization algorithm: {algorithm}"
+        # Export sanity checks
+        if export_config is not None:
+            assert dtype in SUPPORTED_DTYPE, f"Unsupported export dtype: {dtype}"
+        self.torch_dtype = torch_dtype_from_precision(dtype)
+
+    def load_quantizable_model(self, nemo_checkpoint_path: str, calib_tp: int = 1, calib_pp: int = 1) -> llm.GPTModel:
```
Comment (janekl) [inline nemo/collections/llm/quantization/quantizer.py:132]: Could we make it a standalone function in available in some utils?

This will be likely reused soon, for example, in other modelopt workloads:
* https://github.com/NVIDIA/NeMo/blob/main/examples/nlp/language_modeling/megatron_gpt_distillation.py
* https://github.com/NVIDIA/NeMo/blob/main/examples/nlp/language_modeling/megatron_gpt_prune.py

File: nemo/collections/llm/quantization/quantizer.py [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+        # layer implementation from megatron/core/transformer/dot_product_attention.py to be functional.
+        model_cfg.name = "modelopt"
+        model_cfg.apply_rope_fusion = False
+        return model_cfg
+
+    def _get_decoder_type(self, config: llm.GPTConfig):
+        return self.export_config.get("decoder_type", None) or get_modelopt_decoder_type(config)
+
+    @staticmethod
+    def _get_unwrapped_mcore_model(model: llm.GPTModel):
```
Comment (janekl) [inline nemo/collections/llm/quantization/quantizer.py:178]: This function also deserves to be a part of some utils package - I believe it will be useful elsewhere

File: nemo/collections/llm/quantization/quantizer.py [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+        from megatron.core.models.gpt import GPTModel as MCoreGPTModel
+
+        unwrapped_model = model
+        while not isinstance(unwrapped_model, MCoreGPTModel):
+            unwrapped_model = unwrapped_model.module
+
+        return unwrapped_model
+
+    def quantize(self, wrapped_model: llm.GPTConfig, forward_loop):
```
Comment (janekl) [inline nemo/collections/llm/quantization/quantizer.py:187]: Argument `wrapped_model` is of type "config"?

Could we perhaps name it simply `model`?

File: nemo/collections/llm/quantization/quantizer.py [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+
+    def export(self, model: llm.GPTModel, nemo_checkpoint_path: Optional[str] = None) -> None:
+        assert self.export_config is not None, "Export config is not set"
+        # TODO: Add sample generate
+        # TODO: Support NeMo 2:
+        # if model.cfg.megatron_amp_O2:
+        #     model.model = unwrap_model(model.model, Float16Module)
+        export_dir = self.export_config["path"]
+        use_nfs_workspace = (model.trainer._fabric.__io__.num_nodes > 1) or (
+            model.config.pipeline_model_parallel_size > 1
```
Comment (janekl) [inline nemo/collections/llm/quantization/quantizer.py:286]: Does `model.config.pipeline_model_parallel_size > 1` also require using NFS? Previously it was not the case

File: tests/collections/llm/test_hf_import.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    return parser.parse_args()
+
+
+if __name__ == '__main__':
+    args = get_args()
+
+    model = llm.LlamaModel(config=llm.Llama2Config7B)
+    nemo2_path = llm.import_ckpt(model, "hf://" + args.hf_model, output_path=Path(args.output_path))
+
+    trainer = nl.Trainer(
```
Comment (janekl) [inline tests/collections/llm/test_hf_import.py:22]: 1. In `Trainer`: are `devices` and `tensor_model_parallel_size` correctly set? I think they should depend on `args`.
2. Can we make `MegatronMixedPrecision` to use model precision: BF16 or FP16?

File: examples/llm/quantization/ptq.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    parser.add_argument('-bs', '--batch_size', default=64, type=int, help='Calibration batch size')
+    parser.add_argument('-sl', '--seq_len', default=128, type=int, help='Length of the tokenized text')
+    parser.add_argument(
+        '-calib_size', '--calibration_dataset_size', default=512, type=int, help='Size of calibration dataset'
+    )
+    parser.add_argument(
+        '-calib_ds',
+        '--calibration_dataset',
+        default="cnn_dailymail",
+        choices=["wikitext", "cnn_dailymail"],
```
Comment (janekl) [inline examples/llm/quantization/ptq.py:60]: Let's also extend choices to `json` to use it in CI, see https://github.com/NVIDIA/NeMo/blob/c5a2a4d0269afac34405ad59567c67d8c96a6ad3/nemo/collections/llm/quantization/quantizer.py#L323-L326? (Related to another comment)

File: .github/workflows/cicd-main.yml [hunk 10]
```diff
@@ -5450,6 +5450,21 @@ jobs:
...
+    uses: ./.github/workflows/_test_template.yml
+    if: contains(fromJSON(needs.cicd-test-container-setup.outputs.test_to_run), 'L2_NeMo_2_PTQ_Llama2_FP8') || needs.cicd-test-container-setup.outputs.all == 'true'
+    with:
+      RUNNER: self-hosted-azure
+      SCRIPT: |
+        python tests/collections/llm/test_hf_import.py --hf_model /home/TestData/nlp/megatron_llama/llama-ci-hf --output_path /tmp/nemo2_ckpt
+
+        python examples/llm/quantization/ptq.py -nc /tmp/nemo2_ckpt -algo fp8 -out /tmp/nemo2_ptq_engine
```
Comment (janekl) [inline .github/workflows/cicd-main.yml:5462]: Could we possibly use a tiny random dataset here? This can be used https://github.com/NVIDIA/NeMo/blob/main/tests/setup/data/create_sample_jsonl.py.

I'm not sure if in GH CI/CD system we are able to pull the large dataset using `from datasets import load_dataset` here https://github.com/NVIDIA/NeMo/blob/c5a2a4d0269afac34405ad59567c67d8c96a6ad3/nemo/collections/llm/quantization/quantizer.py#L313-L331. cc @ko3n1g

File: examples/llm/quantization/ptq.py [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+        type=str,
+        default="no_quant",
+        choices=["no_quant", "int8", "int8_sq", "fp8", "int4_awq", "w4a8_awq", "int4"],
+        help='TensorRT-Model-Optimizer quantization algorithm',
+    )
+    parser.add_argument(
+        '-awq_bs', '--awq_block_size', type=int, default=128, help='Block size for AWQ quantization algorithms'
+    )
+    parser.add_argument('--sq_alpha', type=float, default=0.5, help='Smooth-Quant alpha parameter')
+    parser.add_argument('--enable_kv_cache', type=bool, help='Enables KV-cache quantization')
```
Comment (janekl) [inline examples/llm/quantization/ptq.py:47]: * There is some helper type parser needed in this case as `--enable_kv_cache False` will result in `args.enable_kv_cache == True` now
* We could maybe make it consistent with other parts of Nemo (needs some examples)

File: nemo/collections/llm/quantization/quantizer.py [hunk 12]
```diff
@@ -0,0 +1,332 @@
...
+    for efficient inference. This can be achieved using NeMo inference containers.
+
+    Available quantization methods are listed in `QUANT_CFG_CHOICES` dictionary above.
+    Please consult Model Optimizer documentation https://nvidia.github.io/TensorRT-Model-Optimizer/ for details.
+
+    Quantization algorithm can also be conveniently set to None to perform only weights export step
+    for TensorRT-LLM deployment. This is useful to getting baseline results for a full-precision model.
+    """
+
+    def __init__(self, quantization_config: dict, export_config: dict):
```
Comment (marcromeyn) [inline nemo/collections/llm/quantization/quantizer.py:97]: We would like to move away from dictionaries everywhere, please turn this into dataclasses.

File: nemo/collections/llm/quantization/quantizer.py [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+        trainer = nl.Trainer(
+            devices=calib_tp * calib_pp,
+            strategy=nl.MegatronStrategy(
+                tensor_model_parallel_size=calib_tp,
+                pipeline_model_parallel_size=calib_pp,
+            ),
+            plugins=nl.MegatronMixedPrecision(precision='16-mixed'),
+        )
+        fabric = trainer.to_fabric()
+        trainer.strategy.setup_environment()
```
Comment (marcromeyn) [inline nemo/collections/llm/quantization/quantizer.py:142]: I think `fabric.launch()` would be better

File: .github/workflows/cicd-main.yml (hunk 1 above)
Comment (Laplasjan107) [reply .github/workflows/cicd-main.yml:4312]: Added

File: nemo/collections/llm/quantization/quantizer.py [hunk 14]
```diff
@@ -0,0 +1,332 @@
...
+        model_cfg.name = "modelopt"
+        model_cfg.apply_rope_fusion = False
+        return model_cfg
+
+    def _get_decoder_type(self, config: llm.GPTConfig):
+        return self.export_config.get("decoder_type", None) or get_modelopt_decoder_type(config)
+
+    @staticmethod
+    def _get_unwrapped_mcore_model(model: llm.GPTModel):
+        from megatron.core.models.gpt import GPTModel as MCoreGPTModel
```
[the remark this answers is not in the document]
Comment (janekl) [reply nemo/collections/llm/quantization/quantizer.py]: Discussed: imports done when required for example here https://github.com/NVIDIA/NeMo/blob/e8a801b58ea64f7ddb61c30e3bb71e966f8c103c/nemo/lightning/megatron_parallel.py#L460 so it's OK

File: nemo/collections/llm/quantization/quantizer.py (hunk 7 above)
Comment (janekl) [reply nemo/collections/llm/quantization/quantizer.py:244]: Discussed offline: there are issues with shared memory-based solution, so we need to use NFS currently

File: .github/workflows/cicd-main.yml (hunk 10 above)
Comment (janekl) [reply .github/workflows/cicd-main.yml:4320]: This will be enabled in a subsequent MR as this tiny dataset results in `np.nan` / `np.inf` stats currently and can't be used

File: tests/collections/llm/test_hf_import.py (hunk 8 above)
Comment (Laplasjan107) [reply tests/collections/llm/test_hf_import.py:22]: Discussed offline: There are issues with distributed fabric import + it's just for creating a nemo2 instance

File: examples/llm/quantization/ptq.py (hunk 9 above)
Comment (Laplasjan107) [reply examples/llm/quantization/ptq.py]: Discussed offline: choice extended, but there are some problems with running ptq with this small dataset

File: nemo/collections/llm/quantization/quantizer.py (hunk 12 above)
Comment (Laplasjan107) [reply nemo/collections/llm/quantization/quantizer.py]: Done, I've added `QuantizationConfig` and `ExportConfig` dataclasses

File: nemo/collections/llm/quantization/quantizer.py (hunk 13 above)
Comment (Laplasjan107) [reply nemo/collections/llm/quantization/quantizer.py]: Changed

File: nemo/collections/llm/quantization/quantizer.py [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+import shutil
+from dataclasses import dataclass
+from typing import Optional, Union
+
+import torch
+import torch.distributed as dist
+from datasets import load_dataset
+
+from nemo.collections import llm
+from nemo.collections.nlp.parts.utils_funcs import torch_dtype_from_precision
```
Comment (marcromeyn) [inline nemo/collections/llm/quantization/quantizer.py:25]: Please move this inline. Importing the nlp collections pulls in a lot of dependencies at import-time, so we move those inline.

File: nemo/collections/llm/quantization/quantizer.py [hunk 16]
```diff
@@ -0,0 +1,284 @@
...
+except (ImportError, ModuleNotFoundError) as e:
+    HAVE_MODELOPT = False
+    HAVE_MODELOPT_ERROR = e
+
+
+SUPPORTED_DTYPE = [16, "16", "bf16"]  # Default precision for non-quantized layers
+
+
+@dataclass
+class QuantizationConfig:
```
Comment (marcromeyn) [inline nemo/collections/llm/quantization/quantizer.py:54]: Please add doc-string

File: nemo/collections/llm/quantization/quantizer.py [hunk 17]
```diff
@@ -0,0 +1,284 @@
...
+    algorithm: Optional[str] = "fp8"  # one of QUANT_CFG_CHOICES keys
+    awq_block_size: int = 128
+    sq_alpha: float = 0.5
+    enable_kv_cache: Optional[bool] = None
+
+
+@dataclass
+class ExportConfig:
```
Comment (marcromeyn) [inline nemo/collections/llm/quantization/quantizer.py:62]: Please add doc-string

**Review by janekl — APPROVED** [review]

**Review by ko3n1g — APPROVED** [review]

**Review by marcromeyn — APPROVED** [review]
